compiler-core/src/ast/typed.rs

```diff
@@ -570,6 +588,20 @@ impl TypedExpr {
         }
     }
 
+    #[must_use]
+    fn is_fn_with_pure_body(&self) -> bool {
+        match self {
+            TypedExpr::Fn { body, .. } => body.iter().all(|s| match s {
+                Statement::Expression(expression) => expression.is_pure_value_constructor(),
+                Statement::Assignment(assignment) => assignment.value.is_pure_value_constructor(),
+                // A use is just a function call under the hood, so we don't ever consider it pure
+                Statement::Use(_) => false,
```

Comment:
I just realised since this is the typed ast `use` cannot appear here. In all other places (like codegen) the compiler panics if it meets a `use`. I can do the same here until <https://github.com/gleam-lang/gleam/pull/3824> is merged, then I could change this